fix #2319 Split emitXxx/tryEmitXxx

simonbasle opened the pull request.

simonbasle commented:
@smaldini as I work on this refinement, I found that there is a lot of inconsistencies on how `FAIL_OVERFLOW` is delivered. Sometimes it propagates an exceptions to subscribers AND returns `FAIL_OVERFLOW`, sometimes it propagates to subscribers but returns `OK`, sometimes it just returns `FAIL_OVERFLOW`... The usage of dropped/discard hooks is also a bit debatable in my opinion.

In this PR I tried to make things more predicable so that `tryEmitXxx` methods usually don't invoke hooks, leaving it to the user to decide what to do with the rejected (value|exception).

As a consequence, `emitXxx` methods DO invoke the hooks with the intention of implementing sane default behavior when the user doesn't want to chose what to do with the `Emission` result (or doesn't want to check it at all).

simonbasle commented:
@reactor/core-team and @smaldini this is probably going to be an extensive PR so don't hesitate to start reviewing despite the draft status. I'll try to make small commits as it progresses.

bsideup commented on reactor-core/src/main/java/reactor/core/publisher/DelegateSinkFluxProcessor.java:
```diff
@@ -47,17 +47,17 @@ public Context currentContext() {
 	@Override
 	public void onNext(IN in) {
-		sink.emitNext(in);
+		sink.tryEmitNext(in);
```
IMO this and many other usages should use the regular emitNext that would fail on overflow. Otherwise we're introducing the same problem as in M2

simonbasle replied:
good catch, indeed and same for `DelegateSinkOneMonoProcessor`

simonbasle replied:
this will actually greatly simplify the PR since most uses in tests will be back to calling `emitXxx`

simonbasle replied:
I've rebased so that a single commit has the introduction of tryEmit BUT correct use of emit in test classes

bsideup commented on reactor-core/src/main/java/reactor/core/publisher/DirectProcessor.java:
```diff
@@ -166,13 +189,33 @@ public Emission emitError(Throwable t) {
-	public Emission emitNext(T t) {
+	public void emitNext(T value) {
+		Emission result = tryEmitNext(value);
+		if (result == Emission.FAIL_OVERFLOW) {
```
nit: could use a `switch` for readability _and_ performance

simonbasle replied:
done

bsideup commented on reactor-core/src/main/java/reactor/core/publisher/DirectProcessor.java:
```diff
@@ -147,14 +158,26 @@ public Emission emitComplete() {
+	@Override
+	public void emitError(Throwable error) {
+		Emission result = tryEmitError(error);
+		if (result == Emission.FAIL_TERMINATED) {
```
while migrating some of r-kafka code, I had the following in mind:
what if r-core adds more "negative" `Emission` cases but I only handle one.

Hence the idea:
Add `Emission#isSuccessful`, so that it can be checked in places like this one.

WDYT?

simonbasle replied:
there's `hasEmitted()` and `hasFailed()`. Maybe I should rename `hasEmitted()` to `hasSucceeded()`? not sure what you mean by "checked in places like this one" though... are you suggesting this piece of code should check something more?

bsideup replied:
Ah, cool, so we can use `hasEmitted()` here (although renaming it to `hasSucceeded()` would be nice).
BTW isn't `hasFailed() == !hasEmitted()`? Why both?

re "can be checked in places like this one":
sorry, I meant this method is a good example where such boolean could be handy, and there are more methods like this one (in this PR but also in r-kafka, for example), hence "the flag that can be checked in places like this one"

simonbasle replied:
yeah, I think it was just added because it reads better. And most likely, `hasFailed()` is going to be the more interesting one to check...

I've opened a separate issue to track these

simonbasle replied:
added a commit for #2329 in the latest rebase (with the `orThrow` split out of the original tryEmit commit)

bsideup replied:
Either I don't see it or don't get it :D 

Anyways, shouldn't it be:
```suggestion
		if (result.hasFailed()) {
```
?

simonbasle replied:
no. for an emitError, cancelled or overflown states don't warrant specific action, whereas an attempt at terminating twice is illegal and usually notified to the `onErrorDropped`

bsideup replied:
ok 👍

bsideup commented on reactor-core/src/main/java/reactor/core/publisher/NextProcessor.java:
```diff
@@ -114,17 +114,41 @@ public O block(@Nullable Duration timeout) {
-	public Emission emitEmpty() {
-		return emitValue(null);
+	public final void onComplete() {
+		//no particular error condition handling for onComplete
```
idea:
```java
@SuppressWarnings("unused")
Emission emission = tryEmitEmpty();
```

simonbasle commented:
With the introduction of #2329, I've reorganized the commits a bit with a first interactive rebase: extracted the `inners[]` parameter change in `Operators` in separate commit, split out `orThrow()` change in #2329-dedicated commit and squashed the two biggest commits so that introducing tryEmit doesn't change most tests.

kept other recent commits for review purposes, but I will squash these before merging.

bsideup commented on reactor-core/src/main/java/reactor/core/publisher/UnicastProcessor.java:
```diff
@@ -280,8 +328,7 @@ public Emission emitNext(T t) {
 			}
-			Operators.onDiscard(t, ctx);
-			onError(ex);
+			onError(ex); //FIXME should this really error ??
```
this one is indeed inconsistent with other usages. `tryEmitNext` should probably not emit any error here

simonbasle replied:
I opened a separate issue, `SerializedManySink` is also guilty of a similar thing

bsideup commented on reactor-core/src/test/java/reactor/core/publisher/FluxPublishOnTest.java:
```diff
@@ -1186,8 +1184,15 @@ public void mapNotifiesOnce() throws InterruptedException {
 		for (int i = 0; i < COUNT; i++) {
-			while (s.emitNext(i).hasFailed() );
+			long busyLoops = 0;
+			while (s.tryEmitNext(i).hasFailed() ) {
```
```suggestion
			while (s.tryEmitNext(i).hasFailed()) {
```

simonbasle replied:
the extra space appears in several other places, I'll fix it

bsideup commented on reactor-core/src/test/java/reactor/core/publisher/FluxPublishOnTest.java:
```diff
@@ -1186,8 +1184,15 @@ public void mapNotifiesOnce() throws InterruptedException {
+			long busyLoops = 0;
+			while (s.tryEmitNext(i).hasFailed() ) {
+				busyLoops++;
+				if (busyLoops % 5000 == 0 && System.currentTimeMillis() - start >= 10_0000) {
```
nit: this whole thing can be replaced with Awaitility, I think

simonbasle replied:
I tried, but it makes the test longer and the tuning of poll interval is complicated, I won't have time to fine tune it.

bsideup commented on reactor-core/src/test/java/reactor/core/publisher/FluxPublishOnTest.java:
```diff
@@ -1226,8 +1231,15 @@ public void mapManyFlushesAllValuesThoroughly() throws InterruptedException {
 		for (int i = 1; i <= items; i++) {
-			while (s.emitNext(String.valueOf(i)).hasFailed() );
+			long busyLoops = 0;
+			while (s.tryEmitNext(String.valueOf(i)).hasFailed() ) {
```
ditto

bsideup commented on reactor-core/src/test/java/reactor/core/publisher/scenarios/FluxSpecTests.java:
```diff
@@ -1103,8 +1103,15 @@ public void collectFromMultipleThread1() throws Exception {
 		Flux.range(1, 1000).subscribe(data -> {
-			while(head.emitNext(data).hasFailed());
+			long busyLoops = 0;
+			while(head.tryEmitNext(data).hasFailed()) {
```
ditto

bsideup commented on reactor-core/src/test/java/reactor/core/publisher/scenarios/FluxTests.java:
```diff
@@ -589,7 +589,13 @@ private void parallelBufferedTimeoutTest(int iterations) throws InterruptedExcep
 
 		for (String i : data) {
-			while (deferred.emitNext(i).hasFailed() );
+			long busyLoops = 0;
```
ditto
